io-sim/src/Control/Monad/IOSim.hs

```diff
@@ -777,6 +779,18 @@ schedule thread@Thread{
                                                (CatchFrame handler k ctl) }
       schedule thread' simstate
 
+    Evaluate expr k -> do
+      mbWHNF <- unsafeIOToST $ try $ evaluate expr
+      case mbWHNF of
+        Left e -> do
+          -- schedule this thread to immediately raise the exception
+          let thread' = thread { threadControl = ThreadControl (Throw e) ctl }
+          schedule thread' simstate
+        Right whnf -> do
+          -- continue with the resulting WHNF
+          let thread' = thread { threadControl = ThreadControl (k whnf) ctl }
+          schedule thread' simstate
```

Comment:
Do these two alternatives recursively call `schedule` correctly? I admit that I mostly copied from similar constructors' cases, without considering all of the context that I'm not altering.

Reply:
This is allright.